WordPress/Classes/ViewRelated/Notifications/Controllers/NotificationsViewController.m

```diff
@@ -471,15 +480,81 @@ - (void)trackAppearedIfNeeded
     self.trackedViewDisplay = YES;
 }
 
-- (void)disableInteractionsForNotification:(Notification *)note
+
+#pragma mark - Undelete Mechanism
+
+- (void)showUndeleteForNote:(Notification *)note onTimeout:(NotificationDetailsDeletionActionBlock)onTimeout
+{
+    // Locate the IndexPath
+    NSIndexPath *indexPath = [self.tableViewHandler.resultsController indexPathForObject:note];
+    if (!indexPath) {
+        return;
+    }
+    
+    // Keep a copy of the Destructive Block
+    [self.notificationIdsWithPendingDeletion addObject:note.objectID];
+    
+    // Update the cell
+    [self.tableView reloadRowsAtIndexPaths:@[ indexPath ] withRowAnimation:UITableViewRowAnimationFade];
+    
+    // Dispatch the Action block
+    dispatch_time_t timeout = dispatch_time(DISPATCH_TIME_NOW, (int64_t)(NotificationsUndoTimeout * NSEC_PER_SEC));
+    dispatch_after(timeout, dispatch_get_main_queue(), ^{
+        [self performPendingDeletionAction:onTimeout noteObjectID:note.objectID];
+    });
+}
+
+- (void)performPendingDeletionAction:(NotificationDetailsDeletionActionBlock)deletionBlock noteObjectID:(NSManagedObjectID *)noteObjectID
+{
+    // Was the Deletion Cancelled?
+    if ([self isNoteMarkedForDeletion:noteObjectID] == false) {
+        return;
+    }
+    
+    // Hide the Notification + Hit the Deletion Block
+    [self signalDeletionInProgress:true noteObjectID:noteObjectID];
+    
+    deletionBlock(^(BOOL success) {
+        // Cleanup
+        [self.notificationIdsWithPendingDeletion removeObject:noteObjectID];
+        
+        // On error, let's unhide the row
+        if (!success) {
+            [self signalDeletionInProgress:false noteObjectID:noteObjectID];
+        }
+    });
+}
+
+- (void)signalDeletionInProgress:(BOOL)deletionInProgress noteObjectID:(NSManagedObjectID *)noteObjectID
 {
-    NSIndexPath *indexPath      = [self.tableViewHandler.resultsController indexPathForObject:note];
+    // ++Paranoid++
+    if (!noteObjectID) {
+        return;
+    }
+    
+    NSManagedObjectContext *context = [[ContextManager sharedInstance] mainContext];
+    Notification *note              = (Notification *)[context existingObjectWithID:noteObjectID error:nil];
```

Comment:
For paranoia, it might be nice to check for and log an error vs just passing nil.